### ISSUE TITLE
Support for multi user containers with podman on kubernetes

### ISSUE THREAD

**Issue by a-username-1** [bottlerocket-os/bottlerocket#2163]
**What I'd like:**
We are running bottlerocket in EKS, and we have a use case to support multi user containers with podman on unprivileged kubernetes containers. It's failing on setting the uid/gid maps. Single user containers run fine with podman on this setup.

**Any alternatives you've considered:**
We have discussed using a different OS on the EKS nodes to get this working, but most of them would require disabling SELinux which we don't want to do.

**Comment by jpculp** [comment]
Hi @a-username-1, thanks for reaching out! We're taking a look at this and will get back to you soon.

**Comment by bcressey** [comment]
Some notes from my initial investigation.

uidMappings and gidMappings are part of the OCI spec, and can be set through [POSIX platform mounts](https://github.com/opencontainers/runtime-spec/blob/main/config.md#posix-platform-mounts) and [user namespace mappings](https://github.com/opencontainers/runtime-spec/blob/main/config-linux.md#user-namespace-mappings).

This spec gets passed to runc as "config.json". There's nothing in containerd-cri or kubelet that would populate those fields, which means by default runc won’t do anything.
 
For a quick proof-of-concept, we can take the output of `ctr oci spec` (which is the default base spec, more or less) and identify the edits we’d want to make so that `runc` does the right thing. The revised spec could be baked into the image so that all containers inherit that setting. That would probably fail because `runc` would try to invoke `newuidmap` to apply the change, which isn’t in Bottlerocket.
 
To get to the point where this actually worked in Bottlerocket, we’d need to:
- [ ] Package shadow-utils (and dependencies) so `newuidmap` is available for `runc`
- [ ] Write an OCI hook that could be inserted into the container launch flow
- [ ] Add a setting e.g. `settings.oci-hooks.delegate-uid-gid-range.enabled = true` to toggle that hook
- [ ] Add `/etc/subuid` + `/etc/subgid` files
 
Ideally the `subuid` and `subgid` files would only have a single user and a large range of UIDs and GIDs, but that might depend on what the right way to set up the OCI spec is, which isn’t clear to me at this stage.

**Comment by jpmcb** [comment]
Beginning the initial discovery work here - I'll first document what other operating systems might be doing to accomplish this with their uid/gid mappings. Then, like Ben laid out, we can start to look at implementing some first party code that mimics what `newuidmap` and `newgidmap` are doing (to avoid importing all of `shadowuitils`).

Pushing this out to v1.11.0 to give some breathing room for the investigative work, determining what needs to happen in bottlerocket, and implementing any new utilities.

**Comment by jpmcb** [comment]
Hi @a-username-1 - Would you be able to provide some yaml I can deploy to EKS to reproduce this issue? This will also be useful when deploying to non-bottlerocket nodes to see what the diff might be. Thanks much!

**Comment by a-username-1** [comment]
Yeah, give me some time, should be able to come up with the config to reproduce this.

**Comment by jpmcb** [comment]
Hi @a-username-1 - touching base here: any luck getting a working test case?

**Comment by jpmcb** [comment]
Hi all - I recently discovered that UID and GID mappings in `runc` only occur [when new user namespaces are created](https://github.com/opencontainers/runc/blob/bc13e3327071f67cf1b1cff6f0f9f9ff4c8a4893/libcontainer/container_linux.go#L2157-L2172). This tracks with additional background information I found on the topic. See this chunk from the Linux man pages for user namespaces:

```
When a user namespace is created, it starts out without a mapping
of user IDs (group IDs) to the parent user namespace.  The
/proc/[pid]/uid_map and /proc/[pid]/gid_map files (available
since Linux 3.5) expose the mappings for user and group IDs
inside the user namespace for the process pid.  These files can
be read to view the mappings in a user namespace and written to
(once) to define the mappings.

...

The initial user namespace has no parent namespace, but, for
consistency, the kernel provides dummy user and group ID mapping
files for this namespace.
```

Ref: https://man7.org/linux/man-pages/man7/user_namespaces.7.html

Creation of user namespaces from within Kubernetes is [partially supported](https://kubernetes.io/blog/2022/10/03/userns-alpha/) as of the 1.25 release. However, it is behind an [alpha feature gate](https://cs.k8s.io/?q=UserNamespacesStatelessPodsSupport&i=nope&files=&excludeFiles=&repos=kubernetes/kubernetes) and the [graduation criteria](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#graduation-criteria) are not yet fully defined, so it is hard to say when it would be available for managed clusters like EKS.

Can you clarify whether the [phase 1](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/127-user-namespaces#phase-1-pods-without-volumes) limitations would work for your use case? In particular, pods would be restricted to only a few volume types where files are not shared with other pods.

**Comment by sworisbreathing** [comment]
hi @bcressey / @jpmcb,

User namespaces appear to have gone GA in [Kubernetes v1.33](https://kubernetes.io/blog/2025/04/25/userns-enabled-by-default/). On bottlerocket (at least with EKS Auto Mode) they aren't enabled by default, but you can enable them by setting the `user.max_user_namespaces` sysctl to something greater than 0.

I've been doing some testing on this in EKS Auto Mode over the past week or so, and here's where I've gotten on it:

1. a DaemonSet with a privileged initContainer (followed by an unprivileged pause container) is able to configure the `user.max_user_namespaces` sysctl and enable user namespaces on the node (note: there are better ways to do this in bottlerocket, but this seems to be the only option for EKS Auto Mode)
2. running pods with `hostUsers: false` works as expected (note: binfmt_misc is still problematic but there are workarounds I won't get into here)
3. `buildah build` (using quay.io/buildah/stable) with `CAP_SYS_ADMIN` and `hostUsers: false` works, but only if the container runs as root. Attempting to run with uid/gid 1000 (the `build` user in the buildah image) fails due to `newgidmap: write to gid_map failed: Operation not permitted` (as well as the equivalent error for uid mapping). I've not been able to make this work by tweaking `/etc/subuid/` or `/etc/subgid` in the image itself (I'm not surprised by this - I just wanted to rule it out as a possible workaround)

Does this help as far as understanding what needs to be done for bottlerocket to support this?
